File: punkc/src/emit.ml
open Bir
open Ast
open Utils

let new_emitter () =
  let context = create_context () in
  let int_type = integer_type context in
  let byte_type = byte_type context in
  let bool_type = boolean_type context in
  let str_type = pointer_type byte_type in
  let void_type = void_type context in
  let mdl = create_module context "punkage" in

  let entry_block =
    let ft = function_type int_type [int_type] in
    let the_function = declare_function "main" ft mdl in
    append_block context "entry" the_function in

  let declare_printf mdl =
    let ft = var_arg_function_type int_type [ pointer_type byte_type ] in
    declare_function "printf" ft mdl in

  let declare_exit mdl =
    let ft = function_type void_type [ int_type ] in
    declare_function "exit" ft mdl in

  object (self)
    val mutable the_module = mdl
    val mutable main_block = entry_block
    val mutable builder = make_builder context
    val mutable current_block = entry_block

    method private get_addr env ((_, lv) as tlv) =
      let { Env.named_values; } = env in
      match lv with
      | Evar (i, _) -> Hashtbl.find named_values i
      | Efield (b, (i, Some f)) ->
        let base = self#get_addr env b in
        assert (i >= 0);
        build_struct_gep base i f builder
      | _ -> self#emit_texp env tlv

    method private switch_block blk =
      let ret = current_block in
      current_block <- blk;
      position_at_end blk builder;
      ret

    method private restore_block blk =
      current_block <- blk;
      position_at_end blk builder

    method private emit_func env name (vmcl, cr, body) =
      let fname =
        match name with
        | Some s -> s
        | None -> Env.new_func_name () in
      begin
        let { Env.named_values; } = env in
        let ft =
          function_type
            (self#emit_con env cr)
            (List.map (fun (v, _, c) -> self#emit_con env c) vmcl) in
        let the_function =
          declare_function fname ft the_module in
        let set_param i a =
          match (Array.of_list vmcl).(i) with
          | ((id, Some n), _, c) ->
            set_param_name n a;
            Hashtbl.add named_values id a;
          | _ -> raise (Error "unnamed param") in
        Array.iteri set_param (params the_function);
        (* Create a new basic block to start insertion into. *)
        let block = append_block context "entry" the_function in
        let parent = self#switch_block block in
        let _ = self#emit_stmt env body in
        ignore (build_ret (undef (self#emit_con env cr)) builder);
        (* Validate the generated code, checking for consistency. *)
        (* Llvm_analysis.assert_valid_function the_function; *)
        self#restore_block parent;
        the_function
      end

    method emit_con env c =
      match c with
      | Cint -> int_type
      | Cbool -> bool_type
      | Cstring -> str_type
      | Cnamed (v, c) -> self#emit_con env c
      | Carray (c, _) -> pointer_type (self#emit_con env c)
      | Cprod (cl, _) ->
        struct_type context (List.map (self#emit_con env) cl)
      | Carrow (cl, cr) -> function_type (self#emit_con env cr)
                             (List.map (self#emit_con env) cl)
      | _ -> raise (Fatal ("unimplemented type emission " ^ (string_of_con c)))

    method emit_texp env (c, e) =
      let { Env.named_values; } = env in
      match e with
      | Evar (id, n) when id >= 0 -> begin
        try
        let v = Hashtbl.find named_values id in
        if Hashtbl.mem env.Env.persistent_set id then
          build_load v ((Env.mangle_name id) ^ "_ld") builder
        else
          v
        with e ->
          Option.value_map n ~default:() ~f:print_string;
          raise e
      end
      | Evar (id, _) when id < 0 ->
        raise (Error "unknown variable name")
      | Eint i -> const_int i
      | Estring s ->
        build_global_stringptr s "string_tmp" builder
      | Ebool b ->
        if b then const_bool true else const_bool false
      | Eop (o, el) ->
        begin match o with
          | Add ->
            let lhs = self#emit_texp env (List.nth el 0) in
            let rhs = self#emit_texp env (List.nth el 1) in
            build_add lhs rhs builder
          | Multiply ->
            let lhs = self#emit_texp env (List.nth el 0) in
            let rhs = self#emit_texp env (List.nth el 1) in
            build_mul lhs rhs builder
          | Minus ->
            let lhs = self#emit_texp env (List.nth el 0) in
            let rhs = self#emit_texp env (List.nth el 1) in
            build_sub lhs rhs builder
          | Equal ->
            let lhs = self#emit_texp env (List.nth el 0) in
            let rhs = self#emit_texp env (List.nth el 1) in
            build_icmp Icmp_eq lhs rhs builder
          | Cprintf ->
            let vl = List.map (self#emit_texp env) el in
            begin match lookup_function "printf" the_module with
              | None -> raise (Fatal "printf should be declared")
              | Some printer ->
                build_call printer (Array.of_list vl) "unit" builder
            end
          | Lt ->
            let lhs = self#emit_texp env (List.nth el 0) in
            let rhs = self#emit_texp env (List.nth el 1) in
            build_icmp Icmp_slt lhs rhs builder
          | Idx ->
            let lhs = self#emit_texp env (List.nth el 0) in
            let rhs = self#emit_texp env (List.nth el 1) in
            let p = build_gep lhs [ rhs ] "idx" builder in
            build_load p "ld" builder
        end
      | Efunc (vmcl, cr, body) ->
        self#emit_func env None (vmcl, cr, body)
      | Efield (b, (i, Some f)) ->
        assert (i >= 0);
        build_extractvalue (self#emit_texp env b) i f builder
      | Earray (c, el) ->
        let res = build_array_alloca
          (self#emit_con env c)
          (const_int (List.length el))
          "array_cns" builder in
        let init_elem i e' =
          let gep =
            build_gep res [ const_int i ] "gep" builder in
          let _ = build_store (self#emit_texp env e') gep builder in
          () in
        List.iteri init_elem el;
        res
      | Econ (Cnamed ((_, Some s), Cprod (cl, _))) ->
        let ty = named_struct_type context s the_module in
        let _ =
          struct_set_body ty (List.map (self#emit_con env) cl) false in
        const_nil ()
      | Ector (Cnamed ((_, Some s), _), sel) ->
        begin match type_by_name s the_module with
          | None -> raise (Error "type not found")
          | Some ty ->
            let elems = struct_element_types ty in
            let start = const_struct context (Array.map undef elems) in
            let (_, res) = List.fold_left
              (fun (i, v) e ->
                 (i + 1,
                  build_insertvalue v (self#emit_texp env e) i "field" builder))
              (0, start) (List.map snd sel) in
            res
        end
      | Eapp (f, params) ->
        build_call (self#emit_texp env f)
          (Array.of_list (List.map (self#emit_texp env) params)) "res" builder
      | _ -> raise (Fatal ("expr code emission unimplemented yet" ^ (string_of_exp e)))

    method emit_stmt env s =
      let { Env.named_values; } = env in
      match s with
      | Sret te ->
        let ret = self#emit_texp env te in
        ignore (build_ret ret builder);
      | Sdecl ((id, n), m, ((c, _) as te)) ->
        let c' = self#emit_con env c in
        if env.is_top then begin
          let env' = { env with is_top = false; } in
          match te with
          | (_, Efunc (vmcl, cr, body)) ->
            let func_name = Env.mangle_func_name id in
            ignore (self#emit_func env' (Some func_name) (vmcl, cr, body));
          | (_, Econ _) ->
            let value = self#emit_texp env' te in
            ignore (define_global (Env.mangle_name id)
                      Bir_unit_type value the_module);
          | _ ->
            Hashtbl.add env.persistent_set id ();
            let addr = define_global (Env.mangle_name id) c' (undef c')
                the_module in
            let value = self#emit_texp env' te in
            Hashtbl.add named_values id addr;
            let parent = self#switch_block main_block in
            ignore (build_store value addr builder);
            self#restore_block parent;
        end else begin
          let value = self#emit_texp env te in
          Hashtbl.add env.persistent_set id ();
          let addr =
            build_alloca
              (self#emit_con env c) (Env.mangle_name id) builder in
          ignore (build_store value addr builder);
          Hashtbl.add named_values id addr;
        end
      | Sblk sl ->
        if env.is_top then begin
          let _ = declare_exit the_module in
          let _ = declare_printf the_module in
          let update_global env s =
            match s with
            | Sdecl ((id, _), _, (c, Efunc _)) ->
              let c' = self#emit_con env c in
              let func = declare_function (Env.mangle_func_name id)
                   c' the_module in
              Hashtbl.add env.persistent_set id ();
              let addr = define_global (Env.mangle_name id)
                  c' func the_module in
              Hashtbl.add named_values id addr;
            | _ -> () in
          List.iter (update_global env) sl;
          List.iter (self#emit_stmt env) sl;
          let _ = build_ret (const_int 0) builder in
          ()
        end else
          List.iter (self#emit_stmt env) sl;
      | Sexpr te ->
        let _ = self#emit_texp env te in ()
      | Sasgn (lval, te) ->
        ignore
          (build_store (self#emit_texp env te)
             (self#get_addr env lval) builder);
      | Sif (te, s0, s1) ->
        let pred = self#emit_texp env te in
        (* Grab the first block so that we might later add the conditional
         * branch to it at the end of the function. *)
        let start_bb = insertion_block builder in
        let the_function = block_parent start_bb in

        let then_bb = append_block context "then" the_function in

        (* Emit 'then' value. *)
        position_at_end then_bb builder;
        self#emit_stmt env s0;

        (* Codegen of 'then' can change the current block, update then_bb for
         * the phi. We create a new name because one is used for the phi node,
         * and the other is used for the conditional branch. *)
        let new_then_bb = insertion_block builder in

        (* Emit 'else' value. *)
        let else_bb = append_block context "else" the_function in
        position_at_end else_bb builder;
        self#emit_stmt env s1;

        (* Codegen of 'else' can change the current block, update else_bb for
         * the phi. *)
        let new_else_bb = insertion_block builder in

        (* Emit merge block. *)
        let merge_bb = append_block context "cont" the_function in

        (* Return to the start block to add the conditional branch. *)
        position_at_end start_bb builder;
        ignore (build_cond_br pred then_bb else_bb builder);

        (* Set a unconditional branch at the end of the 'then' block and the
         * 'else' block to the 'merge' block. *)
        position_at_end new_then_bb builder; ignore (build_br merge_bb builder);
        position_at_end new_else_bb builder; ignore (build_br merge_bb builder);

        (* Finally, set the builder to the end of the merge block. *)
        position_at_end merge_bb builder;

      | Swhile (e, s) ->

        let start_bb = insertion_block builder in
        let the_function = block_parent start_bb in

        let cond_bb = append_block context "cond" the_function in

        position_at_end cond_bb builder;
        let pred = self#emit_texp env e in

        let new_cond_bb = insertion_block builder in

        let loop_bb = append_block context "loop" the_function in
        position_at_end loop_bb builder;
        self#emit_stmt env s;

        let new_loop_bb = insertion_block builder in

        (* Emit merge block. *)
        let merge_bb = append_block context "cont" the_function in

        position_at_end start_bb builder;
        ignore (build_br cond_bb builder);

        position_at_end new_cond_bb builder;
        ignore (build_cond_br pred loop_bb merge_bb builder);

        position_at_end new_loop_bb builder;
        ignore (build_br new_cond_bb builder);

        position_at_end merge_bb builder;

    method get_module () = the_module
  end

File: punkc_fs/backend/emit.fs
module Emit

open Bir
open ir.Ir
open Errors

let valueMap x d f =
    match x with
    | Some x' -> f x'
    | None -> d

type Emitter(mdl, context) =
    let intType = integer_type context
    let byteType = byte_type context
    let boolType = boolean_type context
    let strType = pointer_type byteType
    let unitType = void_type context

    let entryBlock =
        let ft = function_type intType [ intType ]
        let theFunction = declare_function "main" ft mdl
        append_block context "entry" theFunction

    let mutable theModule = mdl
    let mutable mainBlock = entryBlock
    let mutable builder = make_builder context
    let mutable currentBlock = entryBlock

    let declarePrintf mdl =
        let ft = var_arg_function_type intType [ pointer_type byteType ]
        declare_function "printf" ft mdl

    let declareExit mdl =
        let ft = function_type unitType [ intType ]
        declare_function "exit" ft mdl

    let getFuncName i = "func_" + (string i)

    let getGlobalName i = "global_" + (string i)

    member private this.EmitRef (env : Env.env) ((_, v) as tv) =
        match v with
        | Evar(id, _) -> env.named_refs.Item(id)
        | Efield(b, (idx, Some f)) ->
            let b = this.EmitRef env b
            assert (idx >= 0)
            build_struct_gep b idx f builder
        | _ -> raise (BackendFatal "rvalue is referenced")

    member private this.SwitchBlock blk =
        let ret = currentBlock
        currentBlock <- blk
        position_at_end blk builder
        ret

    member private this.RestoreBlock blk =
        currentBlock <- blk
        position_at_end blk builder

    member private this.EmitFunc (env : Env.env) (id, vmcl, cr, body) =
        let (i, _) = id
        let fname = getFuncName i
        let ft =
            function_type (this.EmitCon env cr)
                (List.map (fun (_, c) -> this.EmitCon env c) vmcl)
        let theFunction = declare_function fname ft theModule
        let prologue = append_block context "prologue" theFunction
        let parent = this.SwitchBlock prologue
        let setParam i a =
            match (Array.ofList vmcl).[i] with
            | ((id, Some n), c) ->
                set_param_name n a
                let addr = build_alloca (this.EmitCon env c) n builder
                build_store a addr builder |> ignore
                env.named_refs.Add(id, addr)
            | _ -> raise (BackendError "unnamed param")
        Array.iteri setParam (get_params theFunction)
        (* Create a new basic block to start insertion into. *)
        let block = append_block context "entry" theFunction
        build_br block builder |> ignore
        this.SwitchBlock block |> ignore
        this.EmitStmt env body |> ignore
        ignore (build_ret (undef (this.EmitCon env cr)) builder)
        (* Validate the generated code, checking for consistency. *)
        (* Llvm_analysis.assert_valid_function the_function; *)
        this.RestoreBlock parent
        theFunction

    member this.EmitCon env c =
        match c with
        | Cint -> intType
        | Cbool -> boolType
        | Cstring -> strType
        | Cstruct(id, _) ->
            let stl = Env.lookupStruct env id
            let cl = List.map snd stl
            struct_type context (List.map (this.EmitCon env) cl)
        | Carray c -> pointer_type (this.EmitCon env c)
        | Cprod [] -> intType // FIXME
        | Carrow(cl, cr) ->
            function_type (this.EmitCon env cr)
                (List.map (this.EmitCon env) cl)
        | _ -> raise (BackendFatal("unimplemented type emission " + (string c)))

    member this.EmitTexp (env : Env.env) (c, e) =
        match e with
        | Evar(id, n) when id >= 0 ->
            try
                let v = env.named_refs.Item(id)
                build_load v ((Env.mangleName id) + "_ld") builder
            with e ->
                valueMap n () (printf "%s")
                raise e
        | Evar(id, _) when id < 0 ->
            raise (BackendError "unknown variable name")
        | Eint i -> const_int i
        | Estring s -> build_global_stringptr s "string_tmp" builder
        | Ebool b ->
            if b then const_bool true
            else const_bool false
        | Eop(o, el) ->
            (match o with
             | Add ->
                 let lhs = this.EmitTexp env (List.item 0 el)
                 let rhs = this.EmitTexp env (List.item 1 el)
                 build_add lhs rhs builder
             | Multiply ->
                 let lhs = this.EmitTexp env (List.item 0 el)
                 let rhs = this.EmitTexp env (List.item 1 el)
                 build_mul lhs rhs builder
             | Minus ->
                 let lhs = this.EmitTexp env (List.item 0 el)
                 let rhs = this.EmitTexp env (List.item 1 el)
                 build_sub lhs rhs builder
             | Equal ->
                 let lhs = this.EmitTexp env (List.item 0 el)
                 let rhs = this.EmitTexp env (List.item 1 el)
                 build_icmp Icmp_eq lhs rhs builder
             | Cprintf ->
                 let vl = List.map (this.EmitTexp env) el
                 (match lookup_function "printf" theModule with
                  | None -> raise (BackendFatal "printf should be declared")
                  | Some printer ->
                      build_call printer (Array.ofList vl) "unit" builder)
             | Lt ->
                 let lhs = this.EmitTexp env (List.item 0 el)
                 let rhs = this.EmitTexp env (List.item 1 el)
                 build_icmp Icmp_slt lhs rhs builder
             | Idx ->
                 let lhs = this.EmitTexp env (List.item 0 el)
                 let rhs = this.EmitTexp env (List.item 1 el)
                 let p = build_gep lhs [ rhs ] "idx" builder
                 build_load p "ld" builder)
        | Efunc(id, vmcl, cr, body) -> this.EmitFunc env (id, vmcl, cr, body)
        | Efield(b, (i, Some f)) ->
            assert (i >= 0)
            build_extractvalue (this.EmitTexp env b) i f builder
        | Earray(el) ->
            match c with
            | Carray elem_ty ->
                let c' = this.EmitCon env elem_ty
                let l = const_int (List.length el)
                let res = build_array_alloca c' l "array_cns" builder
                let init_elem i e' =
                    let gep = build_gep res [ const_int i ] "gep" builder
                    let _ = build_store (this.EmitTexp env e') gep builder
                    ()
                List.iteri init_elem el
                res
            | _ -> raise (BackendFatal "TODO Emitter.EmitTexp")
        | Ector(Cstruct(_, Some s), sel) ->
            (match type_by_name s theModule with
             | None -> raise (BackendError "TODO Emitter.EmitTexp")
             | Some ty ->
                 let elems = struct_element_types ty
                 let start = const_struct context (Array.map undef elems)
                 let roll =
                     (fun (i, v) e ->
                     (i + 1,
                      build_insertvalue v (this.EmitTexp env e) i "field"
                          builder))
                 let (_, res) = List.fold roll (0, start) (List.map snd sel)
                 res)
        | Eapp(f, args) ->
            build_call (this.EmitTexp env f)
                (Array.ofList (List.map (this.EmitTexp env) args)) "res"
                builder
        | Etuple [] ->
            const_int 0 // FIXME
        | _ ->
            raise (BackendFatal "TODO Emitter.EmitTexp")

    member this.EmitStmt (env : Env.env) s =
        match s with
        | Sret te ->
            let ret = this.EmitTexp env te
            ignore (build_ret ret builder)
        | Sdecl((id, n), ((c, _) as te)) ->
            let c' = this.EmitCon env c
            let value = this.EmitTexp env te
            // FIXME build_alloca -> alloc_box
            let addr =
                build_alloca (this.EmitCon env c) (Env.mangleName id)
                    builder
            ignore (build_store value addr builder)
            env.named_refs.Add(id, addr)
        | Sblk sl ->
            List.iter (this.EmitStmt env) sl
        | Sexpr te ->
            let _ = this.EmitTexp env te in ()
        | Sasgn(lval, te) ->
            ignore
                (build_store (this.EmitTexp env te) (this.EmitRef env lval)
                     builder)
        | Sif(te, s0, s1) ->
            let pred = this.EmitTexp env te
            (* Grab the first block so that we might later add the conditional
         * branch to it at the end of the function. *)
            let startBB = insertion_block builder
            let theFunction = block_parent startBB
            let thenBB = append_block context "then" theFunction
            (* Emit 'then' value. *)
            position_at_end thenBB builder
            this.EmitStmt env s0
            (* Codegen of 'then' can change the current block, update thenBB for
         * the phi. We create a new name because one is used for the phi node,
         * and the other is used for the conditional branch. *)
            let newThenBB = insertion_block builder
            (* Emit 'else' value. *)
            let elseBB = append_block context "else" theFunction
            position_at_end elseBB builder
            this.EmitStmt env s1
            (* Codegen of 'else' can change the current block, update elseBB for
         * the phi. *)
            let newElseBB = insertion_block builder
            (* Emit merge block. *)
            let mergeBB = append_block context "cont" theFunction
            (* Return to the start block to add the conditional branch. *)
            position_at_end startBB builder
            ignore (build_cond_br pred thenBB elseBB builder)
            (* Set a unconditional branch at the end of the 'then' block and the
         * 'else' block to the 'merge' block. *)
            position_at_end newThenBB builder
            ignore (build_br mergeBB builder)
            position_at_end newElseBB builder
            ignore (build_br mergeBB builder)
            (* Finally, set the builder to the end of the merge block. *)
            position_at_end mergeBB builder
        | Swhile(e, s) ->
            let startBB = insertion_block builder
            let theFunction = block_parent startBB
            let condBB = append_block context "cond" theFunction
            position_at_end condBB builder
            let pred = this.EmitTexp env e
            let new_condBB = insertion_block builder
            let loopBB = append_block context "loop" theFunction
            position_at_end loopBB builder
            this.EmitStmt env s
            let new_loopBB = insertion_block builder
            (* Emit merge block. *)
            let mergeBB = append_block context "cont" theFunction
            position_at_end startBB builder
            ignore (build_br condBB builder)
            position_at_end new_condBB builder
            ignore (build_cond_br pred loopBB mergeBB builder)
            position_at_end new_loopBB builder
            ignore (build_br new_condBB builder)
            position_at_end mergeBB builder

    member private this.DeclFunc env s =
        match s with
        | Sdecl((i, n), (c, Efunc _)) ->
            let t = this.EmitCon env c
            let f = declare_global (getGlobalName i) t theModule
            env.named_refs.Add(i, f)
        | _ -> ()

    member this.ScanHeader(env : Env.env) = function
        | ((i, Some s), Dstruct(_, stl)) ->
            named_struct_type context s theModule |> ignore
            env.struct_def.Add(i, stl)
        | (_, Dstruct _) -> raise (BackendFatal "TODO Emitter.ScanHeader")
        | (_, Diface _) -> raise (BackendFatal "TODO Emitter.ScanHeader")
        | (_, Dalias _) -> raise (BackendFatal "TODO Emitter.ScanHeader")

    member this.EmitHeader(env : Env.env) = function
        | ((_, Some s), Dstruct(_, stl)) ->
            let ty = named_struct_type context s theModule
            let tl = List.map snd stl
            struct_set_body ty (List.map (this.EmitCon env) tl) false |> ignore
        | (_, Dstruct _) -> raise (BackendFatal "TODO Emitter.ScanHeader")
        | (_, Diface _) -> raise (BackendFatal "TODO Emitter.ScanHeader")
        | (_, Dalias _) -> raise (BackendFatal "TODO Emitter.ScanHeader")

    member private this.EmitGlobal (env : Env.env) s =
        match s with
        | Sdecl((id, n), ((c, v) as te)) ->
            let parent = this.SwitchBlock mainBlock
            let value = this.EmitTexp env te
            let t = this.EmitCon env c
            let addr = declare_global (getGlobalName id) t theModule
            build_store value addr builder |> ignore
            this.RestoreBlock parent
            // TODO migrate to TryAdd for dotnet core 3.0
            if not (env.named_refs.ContainsKey(id)) then
                env.named_refs.Add(id, addr)
        | _ ->
            this.EmitStmt env s

    member private this.EmitMainPrologue(env : Env.env) =
        let parent = this.SwitchBlock mainBlock
        build_ret (const_int 0) builder |> ignore
        this.RestoreBlock parent

    member this.EmitModule env (tt, prog) =
        declareExit theModule |> ignore
        declarePrintf theModule |> ignore
        List.iter (this.ScanHeader env) tt
        List.iter (this.EmitHeader env) tt
        List.iter (this.DeclFunc env) prog
        List.iter (this.EmitGlobal env) prog
        this.EmitMainPrologue env

    member this.GetModule() = theModule

let newEmitter() =
    let context = create_context()
    let mdl = create_module context "punkage"
    new Emitter(mdl, context)
